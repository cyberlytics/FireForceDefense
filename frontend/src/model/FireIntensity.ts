enum FireIntensity {
    INTENSITY_0,
    INTENSITY_1,
    INTENSITY_2,
    INTENSITY_3,
    INTENSITY_4,
    INTENSITY_5,
    INTENSITY_6,
    INTENSITY_7,
    INTENSITY_8,
    INTENSITY_9,
    INTENSITY_10,
    INTENSITY_11,
    INTENSITY_12,
}

export default FireIntensity;
