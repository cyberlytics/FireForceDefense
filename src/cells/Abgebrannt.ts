import Cell from '../model/Cell';

export default class Abgebrannt extends Cell {
    readonly description = 'Abgebrannt: Description';
    readonly name = 'Abgebrannt';
    readonly id = 'Abgebrannt';
    readonly groups = ['land'];
    readonly damageMax = +Infinity;
    readonly spreadAmount = Cell.SPREAD_AMOUNT_MEDIUM;
    readonly ignitionChance = 0;
    readonly ignitionThreshold = +Infinity;
    readonly fireGrowAmount = -1;
    readonly maxFireIntensity = Cell.MAX_FIRE_INTENSITY_HIGH;
}
