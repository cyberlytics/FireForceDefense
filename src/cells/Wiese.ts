import Cell from '../model/Cell';

export default class Wiese extends Cell {
    readonly description = 'Meadow: Grassy landscape is important for the livestock. Unfortunately its easily devoured by fire.';
    readonly name = 'Wiese';
    readonly id = 'Wiese';
    readonly groups = ['land'];
    readonly damageMax = Cell.DAMAGE_MAX_MEDIUM;
    readonly spreadAmount = Cell.SPREAD_AMOUNT_MEDIUM;
    readonly ignitionChance = Cell.IGNITION_CHANCE_MEDIUM;
    readonly ignitionThreshold = Cell.IGNITION_THRESHOLD_MEDIUM;
    readonly fireGrowChance = 0.4;
    readonly fireGrowAmount = 2;
    readonly maxFireIntensity = Cell.MAX_FIRE_INTENSITY_MEDIUM;
}
