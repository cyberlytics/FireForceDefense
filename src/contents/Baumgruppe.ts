import Content from '../model/Content';
import ContentCategory from '../model/ContentCategory';
import type Cell from '../model/Cell';

export default class Baumgruppe extends Content {
    readonly description = 'Baumgruppe: Description';
    readonly name = 'Baumgruppe';
    readonly id = 'Baumgruppe';
    readonly buildCosts: null = null;
    readonly damageMax = Content.DAMAGE_MAX_MEDIUM;
    readonly removeCosts: number = 5;
    readonly extinguishRange = 0;
    readonly extinguishRate = 0;
    readonly contentCategory = ContentCategory.CONTENT_NATURE;

    isPlaceableOn(cell: Cell): boolean {
        return cell.groups.includes('land');
    }
}
