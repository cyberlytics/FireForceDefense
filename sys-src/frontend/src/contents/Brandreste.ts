import Content from '@model/Content';
import ContentCategory from '@model/ContentCategory';

export default class Brandreste extends Content {
    readonly description =
        'Fire scene residue: The remains after a burn let merely guess the splendor for former nature and civilization.';
    readonly name = 'Brandreste';
    readonly id = 'Brandreste';
    readonly buildCosts: null = null;
    readonly damageMax = +Infinity;
    readonly fireGrowChance = 0;
    readonly fireGrowAmount = 0;
    readonly removeCosts: number = 5;
    readonly extinguishRange = 0;
    readonly extinguishRate = 0;
    readonly extinguishChance = 0;
    readonly contentCategory = ContentCategory.CONTENT_NATURE;

    isPlaceableOn(): boolean {
        return true;
    }
}
