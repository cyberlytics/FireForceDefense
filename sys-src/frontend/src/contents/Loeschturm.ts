import Content from '@model/Content';
import ContentCategory from '@model/ContentCategory';
import type Cell from '@model/Cell';

export default class Loeschturm extends Content {
    readonly description =
        'Fire-fighting tower: A big tower with a built-in water pump and a raised fire extinguishing range.';
    readonly name = 'Loeschturm';
    readonly id = 'Loeschturm';
    readonly buildCosts: number = 25;
    readonly damageMax = Content.DAMAGE_MAX_LOW;
    readonly fireGrowChance = 0;
    readonly fireGrowAmount = 0;
    readonly removeCosts: number = -15;
    readonly extinguishRange = 2;
    readonly extinguishRate = 1;
    readonly extinguishChance = 0.5;
    readonly contentCategory = ContentCategory.CONTENT_EXTINGUISHER;

    isPlaceableOn(cell: Cell): boolean {
        return cell.groups.includes('land');
    }
}
