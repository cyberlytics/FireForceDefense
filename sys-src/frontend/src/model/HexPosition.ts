/**
 * This class stores hexagonal positions.
 * It uses axial coordinates (q and r) internally, but also exposes the calculated cube coordinates (x, y, z).
 * @see https://www.redblobgames.com/grids/hexagons/
 */
import HexDirection from './HexDirection';
import type AxialCoordinates from './AxialCoordinates';
import type CubeCoordinates from './CubeCoordinates';

export default class HexPosition implements AxialCoordinates, CubeCoordinates {

    public get x() {
        return this.q;
    }
    public get y() {
        return -this.q - this.r;
    }
    public get z() {
        return this.r;
    }

    /**
     * Constructor for hexagonal positions specified by axial coordinates.
     * @param q The (skewed) column index.
     * @param r The row index.
     */
    public constructor(q: number, r: number) {
        this.q = q;
        this.r = r;
    }

    private static readonly directionVectors = {
        [HexDirection.DIRECTION_0]: new HexPosition(+1, 0),
        [HexDirection.DIRECTION_60]: new HexPosition(+1, -1),
        [HexDirection.DIRECTION_120]: new HexPosition(0, -1),
        [HexDirection.DIRECTION_180]: new HexPosition(-1, 0),
        [HexDirection.DIRECTION_240]: new HexPosition(-1, +1),
        [HexDirection.DIRECTION_300]: new HexPosition(0, +1),
    };
    public readonly q: number;
    public readonly r: number;

    public static fromAxial({q, r}: AxialCoordinates) {
        return new HexPosition(q, r);
    }
    public static fromCube({x, y, z}: CubeCoordinates) {
        if (x + y + z !== 0) {
            console.warn('Invalid coordinates supplied! Ignoring y component.');
        }
        return new HexPosition(x, z);
    }

    public static add(p1: HexPosition, p2: HexPosition) {
        return new HexPosition(p1.q + p2.q, p1.r + p2.r);
    }

    public static getDirectionVector(direction: HexDirection) {
        return HexPosition.directionVectors[direction];
    }

    public static getDistance(p1: HexPosition, p2: HexPosition) {
        return Math.max(
            Math.abs(p1.x - p2.x),
            Math.abs(p1.y - p2.y),
            Math.abs(p1.z - p2.z),
        );
    }

    public static getWidth() {
        return Math.sqrt(3);
    }

    public static getIncircleRadiusHorizontalComponent(hexRadius: number = 0, size: number = 1) {
        return (0.5 + 0.75 * hexRadius - (hexRadius % 2) * 0.25) * this.getWidth() * size;
    }

    public static getIncircleRadiusVerticalComponent(hexRadius: number = 0, size: number = 1) {
        const tmp = Math.floor((hexRadius - 1) / 2) * 0.75 + 0.5;
        return (tmp + Math.abs(tmp)) * size;
    }

    public static getIncircleRadius(hexRadius: number = 0, size: number = 1) {
        return Math.sqrt(
            Math.pow(this.getIncircleRadiusHorizontalComponent(hexRadius, size), 2)
            + Math.pow(this.getIncircleRadiusVerticalComponent(hexRadius, size), 2)
        );
    }

    public getCenterX(size: number = 1) {
        return size * (1.732050808 * this.q + 0.8660254038 * this.r);
    }

    public getCenterY(size: number = 1) {
        return size * 1.5 * this.r;
    }

    public getNeighborPosition(direction: HexDirection) {
        return HexPosition.add(this, HexPosition.getDirectionVector(direction));
    }

    public toString() {
        return '(' + this.q + ',' + this.r + ')';
    }

    public getPathString(size: number = 1) {
        const s = parseFloat(size.toString());
        const xOff = s * 0.8660254038;
        const halfSize = s * 0.5;
        return (+xOff) + ',' + (-halfSize) + ' '
            + (+xOff) + ',' + (+halfSize) + ' '
            + (0    ) + ',' + (+s       ) + ' '
            + (-xOff) + ',' + (+halfSize) + ' '
            + (-xOff) + ',' + (-halfSize) + ' '
            + (0    ) + ',' + (-s       );
    }
}
