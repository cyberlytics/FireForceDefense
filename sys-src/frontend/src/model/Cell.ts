import FireIntensity from './FireIntensity';
import type Content from './Content';
import type HexPosition from './HexPosition';
import Fire from './Fire';
import type Explainable from './Explainable';

export default abstract class Cell implements Explainable {
    static readonly DAMAGE_MAX_LOW = 20;
    static readonly DAMAGE_MAX_MEDIUM = 40;
    static readonly DAMAGE_MAX_HIGH = 70;

    static readonly MAX_FIRE_INTENSITY_LOW = FireIntensity.INTENSITY_5;
    static readonly MAX_FIRE_INTENSITY_MEDIUM = FireIntensity.INTENSITY_9;
    static readonly MAX_FIRE_INTENSITY_HIGH = FireIntensity.INTENSITY_12;

    static readonly SPREAD_AMOUNT_LOW = 1;
    static readonly SPREAD_AMOUNT_MEDIUM = 2;
    static readonly SPREAD_AMOUNT_HIGH = 3;

    static readonly IGNITION_CHANCE_LOW = 0.05;
    static readonly IGNITION_CHANCE_MEDIUM = 0.15;
    static readonly IGNITION_CHANCE_HIGH = 0.3;

    static readonly IGNITION_THRESHOLD_LOW = FireIntensity.INTENSITY_2;
    static readonly IGNITION_THRESHOLD_MEDIUM = FireIntensity.INTENSITY_5;
    static readonly IGNITION_THRESHOLD_HIGH = FireIntensity.INTENSITY_8;

    abstract readonly id: string;
    abstract readonly name: string;
    abstract readonly description: string;
    abstract readonly damageMax: number;
    abstract readonly maxFireIntensity: FireIntensity;
    abstract readonly fireGrowChance: number;
    abstract readonly fireGrowAmount: number;
    abstract readonly spreadAmount: number;
    abstract readonly ignitionChance: number;
    abstract readonly ignitionThreshold: number;
    abstract readonly groups: string[];

    content: Content | null = null;
    damage = 0;
    fireIntensity: FireIntensity = FireIntensity.INTENSITY_0;
    neighborSpreadTmp: number;
    stepBeginIntensity: FireIntensity = FireIntensity.INTENSITY_0;
    readonly position: HexPosition;

    constructor(position: HexPosition) {
        this.position = position;
    }

    public applyDamage(): boolean {
        this.damage += Fire.intensityToDamage(this.fireIntensity);
        return this.damage >= this.damageMax;
    }
}
