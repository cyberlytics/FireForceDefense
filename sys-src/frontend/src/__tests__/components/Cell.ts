import { mount } from '@vue/test-utils';
import Cell from '@components/cell.vue';
import Wiese from '@cells/Wiese';
import HexPosition from '@model/HexPosition';
import Baumgruppe from '@contents/Baumgruppe';
import FireIntensity from '@model/FireIntensity';
import Abgebrannt from '@cells/Abgebrannt';
import Brandreste from '@contents/Brandreste';
import Loeschturm from '@contents/Loeschturm';
import Steinbrocken from '@contents/Steinbrocken';

describe('Cell.vue', () => {
    test('empty cell matches snapshot', () => {
        const wrapper = mount(Cell, {
            propsData: {
                cell: new Wiese(new HexPosition(0, 0)),
                size: 60,
            },
        });
        expect(wrapper.element).toMatchSnapshot();
    });

    test('cell with content matches snapshot', () => {
        const cell = new Wiese(new HexPosition(0, 0));
        cell.content = new Baumgruppe();
        const wrapper = mount(Cell, {
            propsData: {
                cell: cell,
                size: 60,
            },
        });
        expect(wrapper.element).toMatchSnapshot();
    });

    test('cell with fire matches snapshot', () => {
        const cell = new Wiese(new HexPosition(0, 0));
        cell.fireIntensity = FireIntensity.INTENSITY_12;
        const wrapper = mount(Cell, {
            propsData: {
                cell: cell,
                size: 60,
            },
        });
        expect(wrapper.element).toMatchSnapshot();
    });

    test('cell of type "Abgebrannt" with fire has smoke', () => {
        const cell = new Abgebrannt(new HexPosition(0, 0));
        cell.fireIntensity = FireIntensity.INTENSITY_3;
        const wrapper = mount(Cell, {
            propsData: {
                cell: cell,
                size: 60,
            },
        });
        expect(wrapper.element).toMatchSnapshot();
    });

    test('cell with fire and "Brandreste" as content has smoke', () => {
        const cell = new Wiese(new HexPosition(0, 0));
        cell.content = new Brandreste();
        cell.fireIntensity = FireIntensity.INTENSITY_3;
        const wrapper = mount(Cell, {
            propsData: {
                cell: cell,
                size: 60,
            },
        });
        expect(wrapper.element).toMatchSnapshot();
    });

    test('cell with damaged content has indicator', () => {
        const cell = new Wiese(new HexPosition(0, 0));
        cell.content = new Loeschturm();
        cell.content.damage++;
        const wrapper = mount(Cell, {
            propsData: {
                cell: cell,
                size: 60,
            },
        });
        expect(wrapper.element).toMatchSnapshot();
    });

    test('cell with damaged, undestroyable content has no indicator', () => {
        const cell = new Wiese(new HexPosition(0, 0));
        cell.content = new Steinbrocken();
        cell.content.damage++;
        const wrapper = mount(Cell, {
            propsData: {
                cell: cell,
                size: 60,
            },
        });
        expect(wrapper.element).toMatchSnapshot();
    });

    test('emit on mouse enter', () => {
        const wrapper = mount(Cell, {
            propsData: {
                cell: new Wiese(new HexPosition(0, 0)),
                size: 60,
            },
        });
        wrapper.find('g').trigger('mouseenter');
        expect(wrapper.emitted()['mouseenter-cell']).toBeTruthy();
    });

    test('emit on mouse leave', () => {
        const wrapper = mount(Cell, {
            propsData: {
                cell: new Wiese(new HexPosition(0, 0)),
                size: 60,
            },
        });
        wrapper.find('g').trigger('mouseleave');
        expect(wrapper.emitted()['mouseleave-cell']).toBeTruthy();
    });
});
