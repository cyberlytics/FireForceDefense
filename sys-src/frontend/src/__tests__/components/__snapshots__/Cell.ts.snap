// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell.vue cell of type "Abgebrannt" with fire has smoke 1`] = `
<g
  transform="translate(0,0)"
>
  <polygon
    fill="url(#cell-Abgebrannt)"
    points="51.961524228,-30 51.961524228,30 0,60 -51.961524228,30 -51.961524228,-30 0,-60"
    stroke="#000"
  />
   
  <!---->
   
  <rect
    fill="url(#smoke)"
    fill-opacity=".4"
    height="73.48469226"
    width="73.48469226"
    x="-36.742346142"
    y="-36.742346142"
  />
   
  <!---->
   
  <rect
    fill="url(#fire-1)"
    height="73.48469226"
    width="73.48469226"
    x="-36.742346142"
    y="-36.742346142"
  />
   
  <!---->
   
  <!---->
</g>
`;

exports[`Cell.vue cell with content matches snapshot 1`] = `
<g
  transform="translate(0,0)"
>
  <polygon
    fill="url(#cell-Wiese)"
    points="51.961524228,-30 51.961524228,30 0,60 -51.961524228,30 -51.961524228,-30 0,-60"
    stroke="#000"
  />
   
  <rect
    fill="url(#content-Baumgruppe)"
    height="73.48469226"
    width="73.48469226"
    x="-36.742346142"
    y="-36.742346142"
  />
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
</g>
`;

exports[`Cell.vue cell with damaged content has indicator 1`] = `
<g
  transform="translate(0,0)"
>
  <polygon
    fill="url(#cell-Wiese)"
    points="51.961524228,-30 51.961524228,30 0,60 -51.961524228,30 -51.961524228,-30 0,-60"
    stroke="#000"
  />
   
  <rect
    fill="url(#content-Loeschturm)"
    height="73.48469226"
    width="73.48469226"
    x="-36.742346142"
    y="-36.742346142"
  />
   
  <!---->
   
  <g>
    <rect
      class="damage-bad"
      height="6"
      width="60"
      x="-30"
      y="-39"
    />
     
    <rect
      class="damage-good"
      height="6"
      width="57"
      x="-30"
      y="-39"
    />
  </g>
   
  <!---->
   
  <!---->
   
  <!---->
</g>
`;

exports[`Cell.vue cell with damaged, undestroyable content has no indicator 1`] = `
<g
  transform="translate(0,0)"
>
  <polygon
    fill="url(#cell-Wiese)"
    points="51.961524228,-30 51.961524228,30 0,60 -51.961524228,30 -51.961524228,-30 0,-60"
    stroke="#000"
  />
   
  <rect
    fill="url(#content-Steinbrocken)"
    height="73.48469226"
    width="73.48469226"
    x="-36.742346142"
    y="-36.742346142"
  />
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
</g>
`;

exports[`Cell.vue cell with fire and "Brandreste" as content has smoke 1`] = `
<g
  transform="translate(0,0)"
>
  <polygon
    fill="url(#cell-Wiese)"
    points="51.961524228,-30 51.961524228,30 0,60 -51.961524228,30 -51.961524228,-30 0,-60"
    stroke="#000"
  />
   
  <rect
    fill="url(#content-Brandreste)"
    height="73.48469226"
    width="73.48469226"
    x="-36.742346142"
    y="-36.742346142"
  />
   
  <rect
    fill="url(#smoke)"
    fill-opacity=".4"
    height="73.48469226"
    width="73.48469226"
    x="-36.742346142"
    y="-36.742346142"
  />
   
  <!---->
   
  <rect
    fill="url(#fire-1)"
    height="73.48469226"
    width="73.48469226"
    x="-36.742346142"
    y="-36.742346142"
  />
   
  <!---->
   
  <!---->
</g>
`;

exports[`Cell.vue cell with fire matches snapshot 1`] = `
<g
  transform="translate(0,0)"
>
  <polygon
    fill="url(#cell-Wiese)"
    points="51.961524228,-30 51.961524228,30 0,60 -51.961524228,30 -51.961524228,-30 0,-60"
    stroke="#000"
  />
   
  <!---->
   
  <!---->
   
  <!---->
   
  <rect
    fill="url(#fire-4)"
    height="73.48469226"
    width="73.48469226"
    x="-36.742346142"
    y="-36.742346142"
  />
   
  <!---->
   
  <!---->
</g>
`;

exports[`Cell.vue empty cell matches snapshot 1`] = `
<g
  transform="translate(0,0)"
>
  <polygon
    fill="url(#cell-Wiese)"
    points="51.961524228,-30 51.961524228,30 0,60 -51.961524228,30 -51.961524228,-30 0,-60"
    stroke="#000"
  />
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
   
  <!---->
</g>
`;
